### DISCUSSION TITLE
How to use vertex_attr_array ?

### DISCUSSION THREAD

**Discussion by rokonio** [gfx-rs/wgpu#1790] [Q&A] [answered]
Hi, I'm having a hard time understanding an error. 
I'm following sotrh's wgpu tutorial, and  I figured out it would be a good idea to use the vertex_attr_array macro.
However I got an error when using it. My code:
```rust
impl Vertex {
    pub fn desc<'a>() -> wgpu::VertexBufferLayout<'a> {
        wgpu::VertexBufferLayout {
            array_stride: std::mem::size_of::<Vertex>() as wgpu::BufferAddress,
            step_mode: wgpu::InputStepMode::Vertex,
            attributes: &wgpu::vertex_attr_array![0 => Float32x3, 1 => Float32x3],
        }
    }
}
```
The error:
```text
   Compiling wgpu_learning v0.1.0 (...\rust\learn_wgpu)
error[E0515]: cannot return value referencing temporary value
  --> src\graphics\vertex_bundle.rs:31:9
   |
31 | /         wgpu::VertexBufferLayout {
32 | |             array_stride: std::mem::size_of::<Vertex>() as wgpu::BufferAddress,
33 | |             step_mode: wgpu::InputStepMode::Vertex,
34 | |             attributes: &wgpu::vertex_attr_array![0 => Float32x3, 1 => Float32x3],
   | |                          -------------------------------------------------------- temporary value created here
35 | |         }
   | |_________^ returns a value referencing data owned by the current function

error: aborting due to previous error

For more information about this error, try `rustc --explain E0515`.
error: could not compile `wgpu_learning`

To learn more, run the command again with --verbose.
```
How am I supposed to use the marcro ? And why does the code fail ?

**Comment by kvark** [comment, marked as answer]
Rust is confused and doesn't realize this macro invocation returns a constant expression.
You can work around by doing this following:
```rust
impl Vertex {
    const ATTRIBS: [wgpu::VertexAttribute; 2] = wgpu::vertex_attr_array![0 => Float32x3, 1 => Float32x3];
    pub fn desc<'a>() -> wgpu::VertexBufferLayout<'a> {
        wgpu::VertexBufferLayout {
            array_stride: std::mem::size_of::<Vertex>() as wgpu::BufferAddress,
            step_mode: wgpu::VertexStepMode::Vertex,
            attributes: &Self::ATTRIBS,
        }
    }
}
```

**Reply by CaesarXInsanium** [reply]
It Works

**Reply by matthewgapp** [reply]
@kvark how would you, instead of using a const and without changing the macro itself, change the lifetime of VertexBufferLayout to 'static as per the tutorial? 

<img width="715" alt="image" src="https://user-images.githubusercontent.com/61894094/195449575-321a4bb6-fab4-48c6-a74f-71b93e8215e6.png">

I tried cargo-cult hacking, passing 'static in various places, but no luck 

<img width="642" alt="image" src="https://user-images.githubusercontent.com/61894094/195449403-e8fa559b-b930-4848-b161-e761f7c102c9.png">

**Reply by tpkee** [reply]
> @kvark how would you, instead of using a const and without changing the macro itself, change the lifetime of VertexBufferLayout to 'static as per the tutorial?
> 
> <img alt="image" width="715" src="https://user-images.githubusercontent.com/61894094/195449575-321a4bb6-fab4-48c6-a74f-71b93e8215e6.png">
> 
> I tried cargo-cult hacking, passing 'static in various places, but no luck
> 
> <img alt="image" width="642" src="https://user-images.githubusercontent.com/61894094/195449403-e8fa559b-b930-4848-b161-e761f7c102c9.png">

```
fn desc () -> wgpu::VertexBufferLayout<'static> {
        wgpu::VertexBufferLayout {
            array_stride: std::mem::size_of::<Vertex>() as wgpu::BufferAddress,
            step_mode: wgpu::VertexStepMode::Vertex,
            attributes: &[ // verbose, could be written in a more concise way with vertex_attr_array! macro
                wgpu::VertexAttribute {
                    offset: 0,
                    shader_location: 0,
                    format: wgpu::VertexFormat::Float32x3,
                },
                wgpu::VertexAttribute {
                    offset: std::mem::size_of::<[f32; 3]>() as wgpu::BufferAddress,
                    shader_location: 1,
                    format: wgpu::VertexFormat::Float32x3,
                }
            ]
        }
    }
```
